ambry-mysql/src/main/java/com/github/ambry/mysql/BatchUpdater.java

```diff
@@ -0,0 +1,153 @@
+ * sent to database in one batch. If it's positive, then when the number of buffered sql commands reaches the
+ * maxBatchSize, this batch would be sent to database.
+ *
+ * Sample usage:
+ * <pre>
+ *   BatchUpdater batch = new BatchUpdater(dataAccessor, "insert into myTable values(?)", "myTable", 100);
+ *   for (String name: names) {
+ *       batch.addUpdateToBatch(statement -> {
+ *           statement.setString(1, name);
+ *       });
+ *   }
+ *   batch.commit();
+ * </pre>
+ */
+public class BatchUpdater {
+  private static final Logger logger = LoggerFactory.getLogger(BatchUpdater.class);
+  private final boolean autoCommit;
+  private final PreparedStatement statement;
+  private final int maxBatchSize;
+  private final MySqlDataAccessor dataAccessor;
+  private final String tableName;
+  private int totalBatchSize = 0;
+  private int currentBatchSize = 0;
+  // The time when the first statement is added
+  private long startTime = 0;
+
+  /**
+   * Constructor to instantiate a {@link BatchUpdater}.
+   * @param dataAccessor The {@link MySqlDataAccessor}.
+   * @param sql The sql command to prepare the statement.
+   * @param tableName The name of the table that update statements are targeting at.
+   * @param maxBatchSize The max batch size.
+   * @throws SQLException
+   */
+  public BatchUpdater(MySqlDataAccessor dataAccessor, String sql, String tableName, int maxBatchSize)
+      throws SQLException {
+    this.dataAccessor = Objects.requireNonNull(dataAccessor, "MySqlDataAccessor is empty");
+    this.tableName = Objects.requireNonNull(tableName, "TableName is empty");
+    if (maxBatchSize < 0) {
+      throw new IllegalArgumentException("MaxBatchSize is not valid: " + maxBatchSize);
+    }
+    this.maxBatchSize = maxBatchSize;
+    try {
+      autoCommit = dataAccessor.getAutoCommmit();
+      dataAccessor.setAutoCommit(false);
+      statement = dataAccessor.getPreparedStatement(sql, true);
+      statement.clearBatch();
+    } catch (SQLException e) {
+      dataAccessor.onException(e, Write);
+      logger.error("Failed to prepare for batch insert on {}", tableName, e);
+      throw e;
+    }
+  }
+
+  /**
+   * Add statement to the batch. Since the sql statement provided to the constructor is a prepared statement,
+   * this method only takes a lambda to supply values to the {@link PreparedStatement}.
+   * @param valueSupplier The consumer to supply the values. The {@link PreparedStatement} is the parameter to this
+   *                      consumer. It can also throws an {@link SQLException}.
+   * @throws SQLException
+   */
+  protected synchronized void addUpdateToBatch(GenericThrowableConsumer<PreparedStatement, SQLException> valueSupplier)
+      throws SQLException {
+    try {
+      if (startTime == 0) {
+        startTime = System.currentTimeMillis();
+      }
+      if (maxBatchSize != 0 && currentBatchSize >= maxBatchSize) {
+        executeBatch();
```

Comment:
Just thinking if it is possible to commit after each batch execution instead of committing one large transaction in the end. Would it require additional logic to avoid rewriting successful batch updates during partial failures.

Reply:
That's a good point.  You can run into issues if the transaction is too large.  Have you tested it with the number of records you expect in this use case?

Reply:
updated